[Idea] Add an Advanced Editor for CMS Categories Description (Entity CMSCategory)

Codencode opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project!

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/#pull-requests

For type and category see:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#type--category
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 9.0.x 
| Description?      | This is a draft implementation of the idea I proposed. In the CMS page editor, a button is added that displays a popup allowing you to select a CMS page. Once the page is selected, a placeholder is added in the editor, which during rendering (in the front-office) is replaced with the link or name of the page. I want to clarify that this is a draft and needs improvement, but I would like to know if this idea could be useful. If it is, I can continue working on it.
| Type?             | new feature
| Category?         | FO / BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | See description
| UI Tests          | 
| Fixed issue or discussion?     | Idea #38108
| Related PRs       | 
| Sponsor company   | @Codencode

Codencode committed:
Aggiunta gestione delle pagine CMS

ps-jarvis commented:
<!-- PR_TABLE_DESCRIPTION_ERROR -->
Hi, thanks for this contribution!

I found some issues with the Pull Request description:

* The `category` should be one of these: `FO`, `BO`, `CO`, `IN`, `WS`, `TE`, `LO`, `ME` or `PM`. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#branch))

Would you mind having a look at it? This will help us understand how interesting your contribution is, thank you very much!
### About linked issues
Please consider opening an issue before submitting a Pull Request:
* If it's a bug fix, it helps maintainers verify that the bug is effectively due to a defect in the code, and that it hasn't been fixed already.
* It can help trigger a discussion about the best implementation path before a single line of code is written.
* It may lead the Core Product team to mark that issue as a priority, further attracting the maintainers' attention.

(Note: this is an automated message, but answering it will reach a real human)

Hlavtox commented:
Hey @Codencode, I like the idea. I opened something very simmilar here some time ago - https://github.com/PrestaShop/PrestaShop/issues/28026

Basically, we could create some kind of "shortcodes" like wordpress has. For linking products, for linking CMS pages, for linking categories.

So, what about creating something like this, but more universal?

Codencode commented:
Hi @Hlavtox,

Yes, the idea actually came to me while thinking about WordPress shortcodes.

In fact, it could be worth considering implementing them in product and category descriptions as well. However, this would introduce some complications: on one hand, the popup would need to allow the selection of the entity to link to (product, category, cms, cmsCategory, manufacturer); on the other hand, the replacement of shortcodes with their respective values would need to happen in multiple places in the front-office.

That said, the idea is excellent because it would solve several issues related to URL changes. A very common scenario is when we develop the site locally, insert content (CMS) with links to other CMS pages, and once the site is uploaded online, we have to manually update the links.

Another issue concerns the links of images inserted through the editor: they are saved as absolute paths, and if the site URL changes, those links become invalid and lead to 404 pages. This is a slightly different aspect but still relevant.

I’ll think about it and keep you updated, as this functionality could really simplify a lot of things!

nicosomb commented on admin-dev/themes/new-theme/js/components/tinymce-editor.js:
```diff
@@ -145,6 +160,58 @@ class TinyMCEEditor {
+              });
+
+              editor.windowManager.open({
+                title: 'Aggiungi link interno',
```
It's better to have english words here

Codencode replied:
@nicosomb,

Yes, yes, of course, the texts need to be in English. This is just a draft (let's say more of a test to understand how to implement everything).
As suggested by @Hlavtox, I'm trying to think more globally, so that I can also include references for products, categories, and manufacturers.

nicosomb reviewed (commented):
I like the idea of shortcodes, even if I hate wysiwyg editor ...

Is that possible to have some perfomance tests here? 
Just to check if time to parse content is not increased so much.

Codencode commented:
> I like the idea of shortcodes, even if I hate wysiwyg editor ...
> 
> Is that possible to have some perfomance tests here? Just to check if time to parse content is not increased so much.



> I like the idea of shortcodes, even if I hate wysiwyg editor ...
> 
> Is that possible to have some perfomance tests here? Just to check if time to parse content is not increased so much.

Sorry, but I didn’t understand what you’re referring to. Maybe if you explain it to me, I can try to figure out how to do it.

Also, keep in mind that I’d like to include this functionality on the product, category, and manufacturer pages as well.

nicosomb commented:
@Codencode I wonder if, with huge text and much "shortcodes", the `replaceCmsTags` method couldn't be a little bit "greedy"

Codencode commented:
@nicosomb Okay, I understand, thank you for the clarification.

We definitely need to take this into account, let's see how the feature evolves and we'll adjust accordingly.

Thank you!

nicosomb commented:
@Codencode can you change your pull request, as a draft one please? 
I did the review as it was a "ready to merge" pull request :D that's why I talked about the english words.

Codencode commented:
@nicosomb Sorry, you're right, I did it!

Codencode commented:
Hello @Hlavtox,

Regarding the discussion about making the functionality more global by adding shortcode management to products, categories, and manufacturers, it has emerged that this change could cause issues with product descriptions.

### Identified Issue
I thought of handling the replacement of shortcodes with their respective values in the `ObjectPresenter::present(`) method, specifically here:
https://github.com/PrestaShop/PrestaShop/blob/9f6fb3918a10fd6cb03d3a4196caf54a87065f9d/src/Adapter/Presenter/Object/ObjectPresenter.php#L45-L62

This way, we centralize shortcode management and, based on the type of entity (product, category, manufacturer), we perform the replacement.

However, this solution only handles shortcodes in the context of normal site usage (e.g., when a user views a page).

### Problematic Scenario
What happens when a module exports product descriptions? In this case, the descriptions would contain unprocessed shortcodes, as the replacement only happens in the presenter and not at the **ObjectModel** level.

Possible Solutions

- Do not implement shortcodes in product descriptions: This is the simplest solution, but it limits functionality.
- Handle shortcode replacement at the ObjectModel level: Introduce a system that allows deciding whether to process shortcodes or not, based on the context (e.g., export or site display).

What do you think?
Do you have any ideas?

Hlavtox commented:
@Codencode You are right, we need to think about XML exports etc. BUT, I think we could get around it by using spans with a data attribute - see my original issue proposal.

Then the description would still work, html stripped or not, because spans have no effect. :-)

Codencode commented:
@Hlavtox, yes, I’ve seen your proposal, and I think it solves the problem.

However, I have some doubts about these two cases:

1. Inserting a link in the HTML to a page (CMS, product, etc.) while specifying a custom "anchor text."
2. Inserting a link to an image uploaded directly from the editor.

For case 1, we could solve this by passing the custom "anchor text" as an attribute of the `<span>`.
For case 2, we might handle it by inserting a `<span>` that clearly indicates it should be an image.

Yes, I think this approach could work.

However, there’s still the issue with exports. When retrieving an entity using `ObjectModel`, the description field will contain shortcodes. This means that whoever develops the export feature might not be able to retrieve the "final" description, i.e., the one visible when browsing the site.

Perhaps we could add a parameter to the `ObjectModel::__construct()`  method to specify whether to convert shortcodes
https://github.com/PrestaShop/PrestaShop/blob/9f6fb3918a10fd6cb03d3a4196caf54a87065f9d/classes/ObjectModel.php#L183-L184
or we could create a dedicated method within the `ObjectModel `class for this conversion.

What do you think?

Codencode commented:
The use of the span with a shortcode has a drawback, which is that it won't be visible in the editor, so I think it might cause issues during management.
For example:
If we have this HTML code: `<p>go to: <span product-id="444"></span> and add to cart</p>`
In the editor view, we will see: `go to and add to cart` since the span is empty.

Codencode commented:
Hello @Hlavtox,
I have implemented a first draft of the idea we discussed.
Here is the admin video (product page):

[admin.webm](https://github.com/user-attachments/assets/8379fb42-0f87-422e-b458-a638b775c048)


This is the front-office video showing how the links are displayed on the various pages:

[front.webm](https://github.com/user-attachments/assets/2a65fd22-f963-49b5-b608-e25a00423783)


In the popup displayed when clicking the new button in the editor, you can search by:
- Product
- Category
- Page
- CMS Page
- Manufacturer

When selecting **Ok**, the system adds the span with the "formatted" attributes as an icon, to indicate that there is an "internal link" there.

There are several things to review, the main ones being:

The search for entities (admin) was implemented by creating the method `shortCodesSearchAction`, added to the controller `PrestaShopBundle\Controller\Admin\Improve\Design\CmsPageController`. For convenience, I placed it in this controller, but it obviously needs to be changed. I think it would be better to create a dedicated controller. Additionally, the search was performed directly in the controller (for practicality), but it should definitely be moved, perhaps to a dedicated service. Do you have any ideas for the name of the controller and the service that should handle the search?

The conversion of shortcodes is handled by the method `Tools::processShortcodes()` (a new method), which is called by various "Presenters."

If you'd like, I can share the test branch where I implemented these changes so you can review it.
What do you think?

Hlavtox commented:
@Codencode Dude, you are awesome. :-) Code will be polished, we also need to check performance, but it looks amazing. Love it. 

Let's ping @kpodemski, @jf-viguier, @marsaldev, @ibahloul-ps - guys what do you think about it? :-)

Codencode commented:
@Hlavtox thank you so much!

Yes, absolutely, the code needs a lot of refinement!
The entity search code must definitely be moved out of the `PrestaShopBundle\Controller\Admin\Improve\Design\CmsPageController` because it doesn’t make sense to have it there—I only put it there temporarily.

Codencode commented:
@Hlavtox I've made some small improvements, but there’s still room for further enhancement.
This is the branch where I developed the code: https://github.com/PrestaShop/PrestaShop/compare/develop...Codencode:PrestaShop:idea-38108-global-refactor-01

In the files, there are some notes in Italian that I will remove later (for now, I’ve left them in).
I moved the search logic in the admin to the `PrestaShopBundle\Controller\Admin\CommonController`
https://github.com/Codencode/PrestaShop/blob/823a7ac5e0b2b56af8407608c91086f3ec30c0c4/src/PrestaShopBundle/Controller/Admin/CommonController.php#L371-L389
but I’m not sure if this is the right place or if a dedicated controller should be created instead.
I also created the `PrestaShopBundle\MultiEntitySearch` class to handle entity searches, but this class might also need to be moved somewhere else.

Then, with the commit **Added shortcode management in the ObjectModel** https://github.com/PrestaShop/PrestaShop/commit/591fc8d4ea677652814cd5a3c2ce953a88837fe1
I added shortcode conversion handling within the `ObjectModel`. I’m not sure if this is the best approach, but I think it could be useful for modules that need to export the catalog and therefore perform shortcode conversion. In this case, they just need to call the `proccessShortCodes()` method.
```php
// Load the product with all available languages
$product = new Product($productId);
// Load the product in a specific language (single language mode)
$product = new Product($productId, false, $langId);

// $product->description now contains the original shortcodes
$product->description;
// Process shortcodes to convert them
$product->proccessShortCodes();
// $product->description now contains the converted shortcodes
$product->description;
```

I’m open to suggestions on how to improve this, and I’m aware that some things need to be modified. However, I preferred to implement them, even if not optimally, as at least it seems to work :).

Codencode commented:
@Hlavtox, @kpodemski, @jf-viguier, @marsaldev, @ibahloul-ps
What do you think? Can I create a PR with this feature?
Thanks!

marsaldev commented:
Sorry @Codencode, I just seen the mention 🤦🏻‍♂️

This could be a really important feature to have 🚀
where are you stuck?

Codencode commented:
Hi @marsaldev,
I’m glad you like the idea.
I put this idea aside for a while because I didn’t think there was much interest, so I’ve shelved it for now.

That said, I created a branch where I developed all the code to handle shortcodes in:
- Products
- Categories
- Manufacturers
- CMS pages
- CMS categories

Here’s the link:
https://github.com/PrestaShop/PrestaShop/commit/f9ed18a085ac2c7c9d38b86fd6bedac90983dc74

It can always be revisited. The code includes comments in Italian that I left for convenience, but they should obviously be removed.

I won’t hide the fact that I’m quite busy right now, both with work and personal commitments, but if the idea gains traction, I’ll definitely move it forward as soon as possible!

Codencode commented:
@Hlavtox @marsaldev @kpodemski @nicosomb 
I've created a PR that includes the functionalities described here, extending them to cover products, categories, CMS pages, CMS categories, and manufacturers. For this reason, I opted to create a new PR.

Here's the PR: https://github.com/PrestaShop/PrestaShop/pull/39165

There are certainly areas that could be improved - I look forward to your feedback.

Thanks!
